changelog.d/15708.feature

```diff
@@ -0,0 +1 @@
+Add Unix Socket support for HTTP Replication Listeners. Document and provide usage instructions for utilizing Unix sockets in Synapse. Contributed by Jason Little.
```

Comment:
Trying this out locally, we also need to update the tracing utilities to take unix sockets into account

```
ProxyResource: Error when proxying request: GET xxx -> [Failure instance: Traceback: <class 'AttributeError'>: 'UNIXAddress' object has no attribute 'host'
/home/eric/.cache/pypoetry/virtualenvs/matrix-synapse-xCtC9ulO-py3.11/lib/python3.11/site-packages/twisted/internet/defer.py:1138:fromCoroutine
/home/eric/.cache/pypoetry/virtualenvs/matrix-synapse-xCtC9ulO-py3.11/lib/python3.11/site-packages/twisted/internet/defer.py:1857:_cancellableInlineCallbacks
/home/eric/.cache/pypoetry/virtualenvs/matrix-synapse-xCtC9ulO-py3.11/lib/python3.11/site-packages/twisted/internet/defer.py:1697:_inlineCallbacks
synapse/synapse/http/server.py:242:wrapped_async_request_handler
--- <exception caught here> ---
synapse/synapse/http/server.py:319:_async_render_wrapper
/usr/lib/python3.11/contextlib.py:137:__enter__
synapse/synapse/logging/opentracing.py:1073:trace_servlet
```

Relevant code: [`synapse/logging/opentracing.py#L1073`](https://github.com/matrix-org/synapse/blob/5e82b07d2c79df9623e5567f935e74395c2e5492/synapse/logging/opentracing.py#L1073)

https://github.com/matrix-org/synapse/blob/5e82b07d2c79df9623e5567f935e74395c2e5492/synapse/logging/opentracing.py#L1073

This specific error might be an outbound federation proxy specific problem since the request is originating from a worker communicating with another worker via unix sockets.

But I see that we use `getClientAddress().host` in a lot of places which seems like it may be a more general problem. The Twisted docs for `getClientAddress()` mention that we should be checking the returned type:

```
        This may not be a network address (e.g., a server listening on
        a UNIX domain socket will cause this to return
        L{UNIXAddress}).  Callers must check the type of the returned
        address.
```

It looks like we already have a utility for this which we could use elsewhere:

https://github.com/matrix-org/synapse/blob/5e82b07d2c79df9623e5567f935e74395c2e5492/synapse/http/site.py#L504-L522

Reply:
Actually can reproduce by curling the unix socket directly:

```sh
$ curl --unix-socket /var/run/main_public.socket -XGET http://test/_matrix/federation/v1/version

->

{"errcode":"M_UNKNOWN","error":"Internal server error"}
```

Error in logs:

```
2023-07-10 16:14:14,007 - synapse.http.server - 133 - ERROR - GET-4 - Failed handle request via 'TransportLayerServer': <XForwardedForRequest at 0x7fcf33d90350 method='GET' uri='/_matrix/federation/v1/version' clientproto='HTTP/1.1' site='unix'>
Traceback (most recent call last):
  File "/home/eric/Documents/github/element/synapse/synapse/http/server.py", line 318, in _async_render_wrapper
    with trace_servlet(request, self._extract_context):
  File "/usr/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/home/eric/Documents/github/element/synapse/synapse/logging/opentracing.py", line 1073, in trace_servlet
    tags.PEER_HOST_IPV6: request.getClientAddress().host,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'UNIXAddress' object has no attribute 'host'
```

Reply:
Generally speaking, the `getClientAddress().host` in most places is ok. Most of those are actually `XForwardedForRequest` which brings in `XForwardedForAddress` which magically takes care of extracting the IP from the request. However this spot does need to get an actual string to add to the `tags` for the tracing span, so let's use that helper function you mentioned. It will at the very least give the span *something* to use.

As you mentioned out-of-band, this does bring along a mypy error:
```log
synapse/logging/opentracing.py:1096: error: Incompatible types in assignment (expression has type "Optional[str]", target has type "str")  [assignment]
```
I'm not sure why this is happening, as both `Type`s haven't really changed(they were both strings after all). Here is the problem spot:
https://github.com/matrix-org/synapse/blob/2328e90fbb65216ff84a08834d3cd99573bccdff/synapse/logging/opentracing.py#L1094-L1097

I'm not an open tracing wizard(nor `Type`ing), and this dives into how `RequestMetrics` work a bit. I'm probably way off base but since the `Type` can be optional, and the only `None` I could fine was if open tracing was disabled, I *think* this is ok to ignore with a `type: ignore[assignment]`? Adding that seems to make mypy happy.

The big question I have is: if that is the case, then why did it never error before?

Reply:
Tried to patch it up in [3789a4f](https://github.com/matrix-org/synapse/pull/15708/commits/3789a4f5d06e3a35a6f43a95b126161f5a7b395e)

Reply:
Looking into this a little more, I used my favorite form of figuring out what is going on(DEBUG logging FTW!!) and(this was checked on `develop`, for reference):
`request.request_metrics.start_context.tag` is almost empty for normal testing, and
`request.request_metrics.start_context` is usually a string like `GET-1` and `PUT-2`. I think this is part of logging context stuff(it looks like it ends up in [`_Sentinel`](https://github.com/matrix-org/synapse/blob/2328e90fbb65216ff84a08834d3cd99573bccdff/synapse/logging/context.py#L220) and [`LoggingContext`](https://github.com/matrix-org/synapse/blob/2328e90fbb65216ff84a08834d3cd99573bccdff/synapse/logging/context.py#L258) where it is set to an empty string, and then never actually used? My same question as before(why did it never error before) still stands though. I'm still leaning towards ok to `ignore` though.

Reply:
Thanks for fixing up the `getClientAddress().host` tracing spot :+1: 

> why did it never error before

Agreed, not sure why it's yelling about it now. Ignoring is a fine option to move forward as it's an existing issue.

> and then never actually used?

In terms of being used, `SynapseTags.REQUEST_TAG` says it's to ["used to distinguish full vs incremental syncs, etc"](https://github.com/matrix-org/synapse/blob/a4243183f0b500f9f30f2d24af19f30a99f65f63/synapse/logging/opentracing.py#L314-L315) but it's set to an empty string when I look at those traces (`SyncRestServlet`, `RoomInitialSyncRestServlet`) even though we supposedly set it here:

https://github.com/matrix-org/synapse/blob/a4243183f0b500f9f30f2d24af19f30a99f65f63/synapse/handlers/sync.py#L348-L357


We also use the [`context.tag` in the metrics](https://github.com/matrix-org/synapse/blob/a4243183f0b500f9f30f2d24af19f30a99f65f63/synapse/http/request_metrics.py#L172-L197) but if I look at the Prometheus data, I don't see any data for `matrix.org` for example. I only see it for one other instance we track which is running Synapse 1.83.0. Has it broken since then?

Seems like something good to track in a separate issue.

Reply:
> Seems like something good to track in a separate issue.

Created https://github.com/matrix-org/synapse/issues/15920 to track this separately :fast_forward: